File: write-api-server-scala/src/main/scala/com/github/j5ik2o/adceet/api/write/MainActor.scala
package com.github.j5ik2o.adceet.api.write

import akka.Done
import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ ActorRef, ActorSystem, Behavior, PostStop }
import akka.actor.{ ClassicActorSystemProvider, CoordinatedShutdown }
import akka.cluster.typed.{ Cluster, SelfUp }
import akka.http.scaladsl.Http
import akka.management.cluster.bootstrap.ClusterBootstrap
import akka.management.scaladsl.AkkaManagement
import com.github.j5ik2o.adceet.api.write.http.Routes
import com.github.j5ik2o.adceet.api.write.http.controller.ThreadController
import kamon.Kamon
import wvlet.log.io.StopWatch
// import kamon.Kamon
import org.slf4j.{ Logger, LoggerFactory }
import wvlet.airframe.{ DISupport, Session }

import scala.concurrent.duration._
import scala.concurrent.{ Await, ExecutionContext, ExecutionContextExecutor, Future }
import scala.util.{ Failure, Success }

object MainActor {
  sealed trait Command

  case object MeUp extends Command

  val logger: Logger = LoggerFactory.getLogger(classOf[MainActor])
}

class MainActor(val session: Session, stopWatch: StopWatch) extends DISupport {

  import MainActor._

  def create(args: Args): Behavior[Command] = {
    Behaviors.setup { ctx =>
      if (args.environment == Environments.Production)
        Kamon.init()

      logger.info(s"[${stopWatch.reportElapsedTime}] create")
      val cluster    = Cluster(ctx.system)
      val selfMember = cluster.selfMember
      logger.info(s"[${stopWatch.reportElapsedTime}] Specified role(s) = ${selfMember.roles.mkString(", ")}")

      logger.info(s"selfMember.roles = ${selfMember.roles}")
      require(selfMember.roles.nonEmpty, "akka.cluster.roles are empty")

      val roleNames = (
        if (selfMember.hasRole(RoleNames.Frontend.toString.toLowerCase))
          Seq(RoleNames.Frontend)
        else Seq.empty
      ) ++ (
        if (selfMember.hasRole(RoleNames.Backend.toString.toLowerCase))
          Seq(RoleNames.Backend)
        else Seq.empty
      )

      logger.info(s"roleNames = $roleNames")

      session.withChildSession(DISettings.mainActor(ctx, roleNames)) { childSession =>
        val config       = ctx.system.settings.config
        val adceetConfig = config.getConfig("adceet")
        val host         = adceetConfig.getString("http.host")
        val port         = adceetConfig.getInt("http.port")
        val terminationHardDeadLine =
          adceetConfig.getDuration("management.http.termination-hard-deadline").toMillis.millis
        val loadBalancerDetachWaitDuration =
          adceetConfig.getDuration("management.http.load-balancer-detach-wait-duration").toMillis.millis

        implicit val system: ActorSystem[_]       = ctx.system
        implicit val ec: ExecutionContextExecutor = ctx.executionContext

        val future = for {
          _ <-
            if (roleNames.contains(RoleNames.Frontend))
              startHttpServer(childSession, ctx.system, host, port, terminationHardDeadLine)
            else
              Future.successful(())
          _ <- startHealthCheckServer(ctx.system, loadBalancerDetachWaitDuration)
        } yield ()

        Await.result(future, 10.seconds)

        val clusterBootstrap: ClusterBootstrap = childSession.build[ClusterBootstrap]
        clusterBootstrap.start()

        childSession.build[ActorRef[SelfUp]]

        Behaviors
          .receiveMessage[Command] { case MeUp =>
            Behaviors.same
          }.receiveSignal { case (_, PostStop) =>
            if (args.environment == Environments.Production)
              Kamon.stop()
            Behaviors.same
          }
      }
    }
  }

  private def startHttpServer(
      session: Session,
      system: ActorSystem[_],
      host: String,
      port: Int,
      terminationHardDeadLine: FiniteDuration
  )(implicit classic: ClassicActorSystemProvider, ec: ExecutionContext): Future[Done] = {
    logger.info(s"[${stopWatch.reportElapsedTime}] startHttpServer: start")

    val threadController: ThreadController = session.build[ThreadController]

    val http = Http()
      .newServerAt(host, port)
      .bind(new Routes(threadController).toRoute)
      .map {
        _.addToCoordinatedShutdown(terminationHardDeadLine)
      }
    http.onComplete {
      case Success(serverBinding) =>
        val address = serverBinding.localAddress
        logger.info(
          s"[${stopWatch.reportElapsedTime}] server bound to http://${address.getHostString}:${address.getPort}"
        )
      case Failure(ex) =>
        logger.error(s"[${stopWatch.reportElapsedTime}] Failed to bind endpoint, terminating system: $ex")
        system.terminate()
    }
    val result = http.map(_ => Done)
    logger.info(s"[${stopWatch.reportElapsedTime}] startHttpServer: finish")
    result
  }

  private def startHealthCheckServer(
      system: ActorSystem[_],
      loadBalancerDetachWaitDuration: FiniteDuration
  )(implicit classic: ClassicActorSystemProvider, ec: ExecutionContext): Future[Done] = {
    logger.info(s"[${stopWatch.reportElapsedTime}] startHealthCheckServer: start")

    val typeName            = "akka-management"
    val coordinatedShutdown = CoordinatedShutdown(system)
    val akkaManagement      = AkkaManagement(system)
    val result = akkaManagement.start().map { _ =>
      logger
        .info(
          s"[${stopWatch.reportElapsedTime}] [$typeName] Server(${akkaManagement.settings.getHttpHostname}:${akkaManagement.settings.getHttpPort}) has started."
        )
      coordinatedShutdown.addTask(CoordinatedShutdown.PhaseBeforeServiceUnbind, "management-terminate") { () =>
        for {
          _ <- akkaManagement.stop()
          _ <- Future {
            logger.info(s"[$typeName] $akkaManagement was terminated")
            logger.info(s"[$typeName] Wait for $loadBalancerDetachWaitDuration to detach from LoadBalancer")
          }
          _ <- akka.pattern.after(loadBalancerDetachWaitDuration) {
            Future {
              logger.info(s"[$typeName] Waiting for disconnection from LoadBalancer is completed.")
            }
          }
        } yield {
          Done
        }
      }
      Done
    }
    logger.info(s"[${stopWatch.reportElapsedTime}] startHealthCheckServer: finish")
    result
  }
}

File: write-api-server-kotlin/src/main/kotlin/com/github/j5ik2o/adceet/api/write/MainActor.kt
package com.github.j5ik2o.adceet.api.write

import akka.Done
import akka.actor.CoordinatedShutdown
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.PostStop
import akka.actor.typed.javadsl.AbstractBehavior
import akka.actor.typed.javadsl.ActorContext
import akka.actor.typed.javadsl.Behaviors
import akka.actor.typed.javadsl.Receive
import akka.cluster.typed.Cluster
import akka.cluster.typed.SelfUp
import akka.http.javadsl.Http
import akka.management.cluster.bootstrap.ClusterBootstrap
import akka.management.javadsl.AkkaManagement
import akka.pattern.Patterns
import com.github.j5ik2o.adceet.api.write.adaptor.http.Routes
import com.github.j5ik2o.adceet.api.write.adaptor.http.controller.ThreadController
import kamon.Kamon
import org.kodein.di.DI
import org.kodein.di.DIAware
import org.kodein.di.factory
import org.kodein.di.instance
import org.kodein.di.on
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import wvlet.log.io.StopWatch
import java.time.Duration
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.function.Supplier

class MainActor constructor(override val di: DI, ctx: ActorContext<Command>) :
    AbstractBehavior<MainActor.Companion.Command>(ctx), DIAware {
    companion object {
        private val logger: Logger = LoggerFactory.getLogger(MainActor.javaClass)

        sealed interface Command

        internal object MeUp : Command

        fun create(di: DI, args: ExampleArgs, stopWatch: StopWatch): Behavior<Command> {
            return Behaviors.setup { ctx ->
                if (args.runModeFor == RunMode.PRODUCTION)
                    Kamon.init()

                logger.info("[${stopWatch.reportElapsedTime()}] create")
                val cluster = Cluster.get(ctx.system)
                val selfMember = cluster.selfMember()
                logger.info("[${stopWatch.reportElapsedTime()}] Specified role(s) = ${selfMember.roles.joinToString(", ")}")

                logger.info("selfMember.roles = ${selfMember.roles}")
                require(selfMember.roles.isNotEmpty()) { "akka.cluster.roles are empty" }

                val roleNames = (if (selfMember.hasRole(RoleName.FRONTEND.toString().lowercase(Locale.getDefault()))) {
                    listOf(RoleName.FRONTEND)
                } else {
                    listOf()
                }) + (if (selfMember.hasRole(RoleName.FRONTEND.toString().lowercase(Locale.getDefault()))) {
                    listOf(RoleName.FRONTEND)
                } else {
                    listOf()
                })
                logger.info("roleNames = $roleNames")


                val config = ctx.system.settings().config()
                val adceetConfig = config.getConfig("adceet")
                val host = adceetConfig.getString("http.host")
                val port = adceetConfig.getInt("http.port")
                val terminationHardDeadLine = adceetConfig.getDuration("management.http.termination-hard-deadline")
                val loadBalancerDetachWaitDuration =
                    adceetConfig.getDuration("management.http.load-balancer-detach-wait-duration")

                val ctxWithArgs = DISettings.CtxWithArgs(ctx, roleNames)

                // アプリケーションの起動
                val appServer = if (roleNames.contains(RoleName.FRONTEND)) {
                        startHttpServer(di, ctxWithArgs, host, port, terminationHardDeadLine)
                    } else {
                        CompletableFuture.completedFuture(Unit)
                    }
                appServer.thenApply {
                    // ヘルスチェックの起動
                    startHealthCheckServer(ctx.system, loadBalancerDetachWaitDuration)
                }.toCompletableFuture().get() // FIXME: タイムアウトを設定する

                // クラスターブートストラップ
                val clusterBootstrap: ClusterBootstrap by di.on(ctxWithArgs).instance()
                clusterBootstrap.start()

                // クラスターにジョインしたことを確認するための子アクターを生成する
                val selfUpRefFactory: (ActorRef<Command>) -> ActorRef<SelfUp> by di.on(ctxWithArgs).factory()
                selfUpRefFactory(ctx.self)

                Behaviors.receive(Command::class.java)
                    .onMessage(MeUp::class.java) {
                        MainActor(di, ctx)
                    }
                    .onSignal(PostStop::class.java) {
                        if (args.runModeFor == RunMode.PRODUCTION)
                            Kamon.stop()
                        Behaviors.same()
                    }.build()
            }
        }

        private fun startHttpServer(
            di: DI,
            ctxWithArgs: DISettings.CtxWithArgs,
            host: String,
            port: Int,
            terminationHardDeadLine: Duration
        ): CompletionStage<Done> {
            val threadController: ThreadController by di.on(ctxWithArgs).instance()

            val http = Http.get(ctxWithArgs.ctx.system).newServerAt(host, port)
                .bind(Routes(threadController).toRoute())
                .thenApply {
                    it.addToCoordinatedShutdown(terminationHardDeadLine, ctxWithArgs.ctx.system)
                }.whenComplete { serverBinding, throwable ->
                    serverBinding?.let {
                        val address = it.localAddress()
                        logger.info("server bound to http://${address.hostString}:${address.port}")
                    } ?: throwable?.let {
                        logger.error("Failed to bind endpoint, terminating system: $it")
                        ctxWithArgs.ctx.system.terminate()
                    }
                }
            return http.thenApply { Done.getInstance() }
        }

        private fun startHealthCheckServer(
            system: ActorSystem<Void>,
            loadBalancerDetachWaitDuration: Duration
        ): CompletionStage<Done> {
            val typeName = "akka-management"
            val coordinatedShutdown = CoordinatedShutdown.get(system)
            val akkaManagement = AkkaManagement.get(system)
            return akkaManagement.start().thenApply {
                logger.info("[$typeName] サーバ(${akkaManagement.settings().httpHostname}:${akkaManagement.settings().httpPort})を起動しました")
                coordinatedShutdown.addTask(
                    CoordinatedShutdown.PhaseBeforeServiceUnbind(), "management-terminate",
                    Supplier {
                        logger.info("[$typeName] 終了処理のため $akkaManagement を終了します")
                        akkaManagement.stop().thenApply {
                            logger.info("[$typeName] $akkaManagement を terminate しました")
                            logger.info("[$typeName] LoadBalancer からの切り離しを待つため、$loadBalancerDetachWaitDuration 待機します")
                            Patterns.after(loadBalancerDetachWaitDuration, system) {
                                logger.info("[$typeName] LoadBalancer からの切り離し待ち待機を完了しました")
                                CompletableFuture.completedStage(Done.getInstance())
                            }
                            Done.getInstance()
                        }
                    }
                )
                Done.getInstance()
            }
        }
    }

    override fun createReceive(): Receive<Command> {
        val builder = newReceiveBuilder()
        return builder.build()
    }
}
